classes/Authentication/SAML/XDSamlAuthentication.php

```diff
@@ -110,124 +110,89 @@ public function isSamlConfigured()
+            // information we'll need to do so.
+            $emailAddress = isset($samlAttrs['email_address']) ? $samlAttrs['email_address'][0] : NO_EMAIL_ADDRESS_SET;
+            $systemUserName = isset($samlAttrs['system_username']) ? $samlAttrs['system_username'][0] : $userName;
+            $firstName = isset($samlAttrs['first_name']) ? $samlAttrs['first_name'][0] : 'UNKNOWN';
+            $middleName = isset($samlAttrs['middle_name']) ? $samlAttrs['middle_name'][0] : null;
+            $lastName = isset($samlAttrs['last_name']) ? $samlAttrs['last_name'][0] : null;
+            $personId = \DataWarehouse::getPersonIdFromPII($systemUserName, $samlAttrs['organization'][0]);
+
+            // Attempt to identify which organization this user should be associated with. Prefer
+            // using the personId if not unknown, then fall back to the saml attributes if the
+            // 'organization' property is present, and finally defaulting to the Unknown organization
+            // if none of the preceding conditions are met.
             $userOrganization = $this->getOrganizationId($samlAttrs, $personId);
 
-            if (!isset($samlAttrs["first_name"])) {
-                $samlAttrs["first_name"] = array("UNKNOWN");
-            }
-            if (!isset($samlAttrs["middle_name"])) {
-                $samlAttrs["middle_name"] = array(null);
-            }
-            if (!isset($samlAttrs["last_name"])) {
-                $samlAttrs["last_name"] = array("UNKNOWN");
-            }
             try {
                 $newUser = new \XDUser(
-                    $thisUserName,
+                    $userName,
                     null,
                     $emailAddress,
-                    $samlAttrs["first_name"][0],
-                    $samlAttrs["middle_name"][0],
-                    $samlAttrs["last_name"][0],
+                    $firstName,
+                    $middleName,
+                    $lastName,
                     array(ROLE_ID_USER),
                     ROLE_ID_USER,
                     $userOrganization,
-                    $personId
+                    $personId,
+                    $samlAttrs
                 );
             } catch (Exception $e) {
-                return "EXISTS";
+                throw new Exception('An account is currently configured with this information, please contact an administrator.');
             }
+
             $newUser->setUserType(SSO_USER_TYPE);
+
             try {
                 $newUser->saveUser();
             } catch (Exception $e) {
                 $this->logger->err('User creation failed: ' . $e->getMessage());
-                return false;
+                throw $e;
             }
 
             $this->handleNotifications($newUser, $samlAttrs, ($personId != UNKNOWN_USER_TYPE));
 
             return $newUser;
         }
-        return false;
+        throw new Exception('An unknown error has occurred.');
```

Comment:
```suggestion
        throw new \DataWarehouse\Query\Exceptions\AccessDeniedException('Authentication failure.');
```

It is not really an unknown error. This case runs when the user did not authenticate with the saml provider.